### ISSUE TITLE
can't get fiberdump of stuck tests on OSX

### ISSUE THREAD

**Issue by domdorn** [zio/zio#9089]
Hi!

I'm trying to get a `Fiber.dump` of stuck tests on OSX. 
From `ZIOApp.scala` I got, that this should be possible by sending a `INFO` or `USR1` signal to the java process. 

Somewhere I read that a CTRL+T should do the same. However, this is not working for me :/

I'm running the tests from sbt and have `Test / fork := true` set. 

When I press `CTRL+T`, I just get outputs like these 
```
load: 6.04  cmd: java 66310 waiting 57.48u 5.84s
load: 5.73  cmd: java 66310 waiting 57.59u 5.88s
load: 6.09  cmd: java 66310 waiting 61.76u 7.33s
load: 5.83  cmd: java 66310 waiting 62.55u 7.49s
```

When I do a `kill -SIGINFO 66310` nothing happens. 
When I do a `kill -SIGUSR1 66310`, instead of getting a fiber dump, the whole test is aborted like this:

```
load: 6.09  cmd: java 66310 waiting 61.76u 7.33s
load: 5.83  cmd: java 66310 waiting 62.55u 7.49s
[info] Completed tests
[info] Completed tests
[error] Error during tests:
[error]         Running java with options -classpath /Users/domdorn/acme/calculation-service-demo/target/test-classes:...all the jars on my classpath...:/Users/domdorn/.sbt/boot/scala-2.12.19/org.scala-sbt/sbt/1.10.1/test-agent-1.10.1.jar:/Users/domdorn/.sbt/boot/scala-2.12.19/org.scala-sbt/sbt/1.10.1/test-interface-1.0.jar sbt.ForkMain 63226 failed with exit code 158
[error] (Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 1371 s (22:51), completed 8 Aug 2024, 23:47:50
sbt:calculation-service-demo> 
```

Versions:
SBT: `1.10.1`
ZIO: `2.1.7`
Java: 
```
java 21.0.1 2023-10-17
Java(TM) SE Runtime Environment Oracle GraalVM 21.0.1+12.1 (build 21.0.1+12-jvmci-23.1-b19)
Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 21.0.1+12.1 (build 21.0.1+12-jvmci-23.1-b19, mixed mode, sharing)
```
macOs: `14.2.1 (23C71)`

What am I doing wrong?

**Comment by domdorn** [comment]
[thread_dump.txt](https://github.com/user-attachments/files/16553766/thread_dump.txt)
attached the thread dump, not sure if its helping in any way... sending kill -3 PID created it.

**Comment by kyri-petrou** [comment]
I know this isn't necessarily answering the question, but for debugging stuck tests have you tried `TestAspect.diagnose`?

**Comment by domdorn** [comment]
The problem is, I don't even know which test is the one that is stuck, as sbt usually runs them in parallel and all the threads from the thread dump don't reference any class of my codebase :/ I'll try to add `TestAspect.diagnose` here and there and hope to figure out whats the problem..

**Comment by kyri-petrou** [comment]
Do you know which file is it in? If you do, you can annotate the entire `suite("foo")(...) @@ TestAspect.diagnose(???)`

If you don't, what you can do is define a base trait and have all your test suites extend it. You can then have aspects applied to all tests, e.g., https://github.com/zio/zio/blob/b1dafc5b811087e69e5944db9eb0bb207b03e2e4/test-tests/shared/src/test/scala/zio/test/ZIOBaseSpec.scala#L5-L9

**Comment by jdegoes** [comment]
/bounty $250 for a fix on Mac OS X, if possible.

**Comment by kyri-petrou** [comment]
/attempt #9089

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/kyri-petrou) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @kyri-petrou | &nbsp;&nbsp;&nbsp;31 ZIO bounties <br/> + 2 bounties from 1 project | <div align="center">Scala, Python, <br />Shell & more</div> | <div align="center">﹟9093</div> | [Cancel attempt](https://console.algora.io/api/bounties/cm0mlf3cq000zky03ffwm1jjo/cancel-attempt) |

**Comment by kyri-petrou** [comment]
Seems that the issue is not with MacOS alone but rather with `Test / fork := true`. Setting it to `Test / fork := false` makes it work as expected.

It's unclear what's not working correctly at the moment; does this issue exists only on MacOS? (I don't have a Linux machine to test this unfortunately). I believe that the OS blocks the console output from the process because it's not the main one running in the terminal (the main one is the SBT one which has a different PID). I think that this is why `kill -3` works because that one outputs the thread dump to a file

**Comment by domdorn** [comment]
When `Test / fork := true` is set, sbt creates a own JVM for the tests where this should work as expected, but does not.  
When `Test / fork := false`, the tests run in sbt, wheras sbt (at least on my mac) intercepts the signal and kills itself. 

I looked again at `ZIOApp.scala` and I suspect the check 
```scala
      if (!ZIOApp.installedSignals.getAndSet(true)) {
```
is not doing what we are expecting when we run multiple test suites, as only one (the first suite) is installing signal handlers which is highly likely not the one that is "hanging" ?

**Comment by kyri-petrou** [comment]
Aghhh my bad, it seems there was some rogue ZIO app running in SBT when I tested it and I was getting the fiber dump from it instead. I've managed to reproduce it now

> as only one (the first suite) is installing signal handlers which is highly likely not the one that is "hanging"

I reproduced the issue on an app with just 1 test so I don't think that's what's causing it. My guess at the moment is that it's because we're not running tests as a ZIO app, but we rather convert the effects to futures and run them in the SBT runners. I'll see if we can add the signal handlers when we init the SBT runner
